crates/fs-utils/src/lockfile.rs

```diff
@@ -0,0 +1,91 @@
+use crate::create_parent_dir;
+use std::path::{Path, PathBuf};
+
+#[derive(thiserror::Error, Debug)]
+pub enum LockfileError {
+    #[error("Failed to acquire lock on {file_path:?}: failed to create lockfile {lock_path:?}: {cause}")]
+    Acquire {
+        file_path: PathBuf,
+        lock_path: PathBuf,
+        #[source]
+        cause: std::io::Error,
+    },
+    #[error("Failed to release lock: failed to delete lockfile {lock_path:?}: {cause}")]
+    Release {
+        lock_path: PathBuf,
+        #[source]
+        cause: std::io::Error,
+    },
+}
+
+#[derive(Debug)]
+/// A file used as an exclusive lock on access to another file.
+///
+/// Constructing a `Lockfile` creates the `path` with [`std::fs::File::create_new`],
+/// a.k.a. `O_EXCL`, erroring if the file already exists.
+///
+/// Dropping a `Lockfile` deletes the `path`, releasing the lock.
+///
+/// Used to guarantee exclusive access to the system config file,
+/// in order to prevent racy concurrent modifications.
+pub struct Lockfile {
+    path: PathBuf,
+}
+
+impl Lockfile {
+    /// Acquire an exclusive lock on the file `file_path`.
+    ///
+    /// `file_path` should be the full path of the file to which to acquire exclusive access.
+    pub fn for_file(file_path: &Path) -> Result<Self, LockfileError> {
+        let path = Self::lock_path(file_path);
+
+        // Ensure the directory exists before attempting to create the lockfile.
+        create_parent_dir(file_path).map_err(|cause| LockfileError::Acquire {
+            lock_path: path.clone(),
+            file_path: file_path.to_path_buf(),
+            cause,
+        })?;
+
+        // Open with `create_new`, which fails if the file already exists.
+        std::fs::File::create_new(&path).map_err(|cause| LockfileError::Acquire {
+            lock_path: path.clone(),
+            file_path: file_path.to_path_buf(),
+            cause,
+        })?;
```

Comment:
I don't like the `and_then` formulation because we're doing `>>` here, not `>>=`. I did deduplicate the closure by binding it to a local, but I retained the structure of two different `foo().map_err(bar)?;` calls.